// @famililook/shared/src/upload/validateFile.js
/**
 * File validation for photo uploads.
 * Enforces type whitelist and size ceiling before any processing.
 */

export const ALLOWED_IMAGE_TYPES = [
  'image/jpeg',
  'image/png',
  'image/webp',
  'image/jpg',
]

export const MAX_FILE_SIZE_MB = 10
export const MAX_FILE_SIZE_BYTES = MAX_FILE_SIZE_MB * 1024 * 1024

/**
 * Validate a file for photo upload.
 * @param {File} file
 * @returns {{ valid: boolean, error: string|null }}
 */
export function validateFile(file) {
  if (!file) {
    return { valid: false, error: 'No file provided' }
  }

  if (!ALLOWED_IMAGE_TYPES.includes(file.type)) {
    return {
      valid: false,
      error: `${file.name || 'File'}: unsupported format. Use JPEG, PNG, or WebP.`,
    }
  }

  if (file.size > MAX_FILE_SIZE_BYTES) {
    const sizeMB = (file.size / 1024 / 1024).toFixed(1)
    return {
      valid: false,
      error: `${file.name || 'File'}: ${sizeMB}MB exceeds the ${MAX_FILE_SIZE_MB}MB limit.`,
    }
  }

  return { valid: true, error: null }
}
