// @famililook/shared/src/theme/colors.js
// famililook-shared/src/theme/colors.js
// Design tokens extracted verbatim from famililook-desktop2/src/theme/colors.js + per-product palettes + frozen chemistry contract colours.

// ──────────────────────────────────────────────────────────────
// Base palette (28 tokens) — identical to desktop2
// ──────────────────────────────────────────────────────────────
export const base = {
  // Backgrounds
  bgMain:     '#0D0F1A',
  bgBase:     '#0D0F1A',
  bgCard:     '#161828',
  bgElevated: '#1E2235',
  bgSheet:    '#1E2235',
  bgInput:    '#FFFFFF',

  // Borders
  borderSoft:   'rgba(255,255,255,0.06)',
  borderMedium: 'rgba(255,255,255,0.10)',
  borderFocus:  'rgba(124,58,237,0.4)',

  // Accents
  accentPrimary:      '#7C3AED',
  accentPrimaryLight: '#A78BFA',
  accentPrimaryDark:  '#6D28D9',
  accentSecondary:    '#4F46E5',
  accentTertiary:     '#10B981',
  accentSuccess:      '#10B981',
  accentWarning:      '#F5A623',
  accentError:        '#EF4444',

  // Text
  textPrimary:   '#E8EAF6',
  textSecondary: '#B8BDD6',
  textMuted:     '#8B92A8',
  textSubtle:    '#7078A0',
  textDisabled:  '#3A3F5A',

  // Overlays and gradients
  overlayDark:     'rgba(0,0,0,0.7)',
  overlayLight:    'rgba(255,255,255,0.04)',
  shimmer:         'linear-gradient(90deg, transparent, rgba(255,255,255,0.06), transparent)',
  gradientPrimary: 'linear-gradient(135deg, #7C3AED 0%, #4F46E5 100%)',
  gradientCard:    'linear-gradient(180deg, #1E2235 0%, #161828 100%)',
  gradientHero:    'linear-gradient(180deg, rgba(124,58,237,0.12) 0%, transparent 50%)',
}

// Legacy alias — matches existing desktop2 export name for zero-rename migration in consumer repos.
export const colors = base

// ──────────────────────────────────────────────────────────────
// Shadows (5 tokens)
// ──────────────────────────────────────────────────────────────
export const shadows = {
  sm:   '0 1px 2px rgba(0,0,0,0.3)',
  md:   '0 4px 12px rgba(0,0,0,0.4)',
  lg:   '0 8px 24px rgba(0,0,0,0.5)',
  xl:   '0 16px 48px rgba(0,0,0,0.6)',
  glow: '0 0 20px rgba(124,58,237,0.3)',
}

// ──────────────────────────────────────────────────────────────
// Spacing (8 tokens)
// ──────────────────────────────────────────────────────────────
export const spacing = {
  xs:         '4px',
  sm:         '8px',
  md:         '16px',
  lg:         '24px',
  xl:         '32px',
  xxl:        '48px',
  safeBottom: '34px',
  safeTop:    '44px',
}

// ──────────────────────────────────────────────────────────────
// Typography (7 tokens)
// ──────────────────────────────────────────────────────────────
export const typography = {
  h1:        { fontSize: '28px', fontWeight: 700, lineHeight: 1.2 },
  h2:        { fontSize: '22px', fontWeight: 600, lineHeight: 1.3 },
  h3:        { fontSize: '18px', fontWeight: 600, lineHeight: 1.4 },
  body:      { fontSize: '16px', fontWeight: 400, lineHeight: 1.5 },
  bodySmall: { fontSize: '14px', fontWeight: 400, lineHeight: 1.5 },
  caption:   { fontSize: '12px', fontWeight: 400, lineHeight: 1.4 },
  label:     { fontSize: '13px', fontWeight: 500, lineHeight: 1.3, letterSpacing: '0.02em' },
}

// ──────────────────────────────────────────────────────────────
// Radius (5 tokens)
// ──────────────────────────────────────────────────────────────
export const radius = {
  sm:   '8px',
  md:   '12px',
  lg:   '16px',
  xl:   '24px',
  full: '9999px',
}

// ──────────────────────────────────────────────────────────────
// Per-product accent palettes
// Keys: all lowercase single words — famililook, familimatch, familiuno, familipoker
// ──────────────────────────────────────────────────────────────
export const products = {
  famililook: {
    primary:   '#7C3AED',
    secondary: '#4F46E5',
  },
  familimatch: {
    primary:   '#0A84FF',
    secondary: '#5E5CE6',
  },
  familiuno: {
    primary:   '#30D158',
    secondary: '#0A84FF',
  },
  familipoker: {
    primary:   '#BF5AF2',
    secondary: '#FF375F',
  },
}

// ──────────────────────────────────────────────────────────────
// Chemistry label colours — FROZEN (compare_faces.v1 contract)
// Single source of truth. Never duplicate in product repos.
// ──────────────────────────────────────────────────────────────
export const chemistry = {
  featureTwins:        '#FFD700',
  magneticMatch:       '#8B5CF6',
  complementaryPair:   '#3B82F6',
  interestingContrast: '#14B8A6',
  oppositesAttract:    '#F97316',
}

// Map from backend chemistry_label string to hex
export const chemistryColorMap = {
  'Feature Twins':        chemistry.featureTwins,
  'Magnetic Match':       chemistry.magneticMatch,
  'Complementary Pair':   chemistry.complementaryPair,
  'Interesting Contrast': chemistry.interestingContrast,
  'Opposites Attract':    chemistry.oppositesAttract,
}
