// @famililook/shared/src/theme/dopamine.js
// famililook-shared/src/theme/dopamine.js
// Reward colour vocabulary (dopamine-response research) + confetti palettes. Re-exports chemistry contract colours under renamed keys to avoid barrel collision with colors.js.

import { chemistry, chemistryColorMap, products } from './colors.js'

// ──────────────────────────────────────────────────────────────
// Reward colour vocabulary
// Use these for reward moments, never for structural UI.
// ──────────────────────────────────────────────────────────────
export const reward = {
  // Electric blue — trust, reward confirmation, high scores.
  // Derived from products.familimatch.primary (single source of truth).
  electric: products.familimatch.primary,

  // Violet — anticipation, loading states, "something is coming"
  anticipation: '#7C3AED',

  // Gold — achievement, rare stat, top milestone
  achievement: '#F5A623',

  // Pink — social bonding, shared results, invite moments
  social: '#EC4899',
}

// ──────────────────────────────────────────────────────────────
// Chemistry label colours re-exported under renamed keys.
// Intentional rename avoids `export *` collision in index.js barrel.
// Consumers inside reward components should prefer these names.
// The canonical frozen names (`chemistry`, `chemistryColorMap`) remain
// the source of truth in colors.js.
// ──────────────────────────────────────────────────────────────
export const rewardChemistry = chemistry
export const rewardChemistryColorMap = chemistryColorMap

// ──────────────────────────────────────────────────────────────
// Confetti palettes (used by CelebrationBurst.jsx)
// Keyed chemistry palettes use the exact backend chemistry_label strings.
// All hex values uppercase per shared-package convention.
// ──────────────────────────────────────────────────────────────
export const confetti = {
  // Default high-energy palette
  standard: ['#0A84FF', '#5E5CE6', '#EC4899', '#A78BFA', '#FBBF24'],

  // Achievement palette (gold-dominant)
  achievement: ['#F5A623', '#FFD700', '#FFF3B0', '#EC4899', '#7C3AED'],

  // Chemistry-specific palettes (keyed by backend chemistry_label)
  chemistry: {
    'Feature Twins':        ['#FFD700', '#FFF3B0', '#F5A623', '#FFF', '#0A84FF'],
    'Magnetic Match':       ['#8B5CF6', '#A78BFA', '#EC4899', '#FFF', '#0A84FF'],
    'Complementary Pair':   ['#3B82F6', '#60A5FA', '#A78BFA', '#FFF', '#EC4899'],
    'Interesting Contrast': ['#14B8A6', '#2DD4BF', '#0A84FF', '#FFF', '#F5A623'],
    'Opposites Attract':    ['#F97316', '#FB923C', '#EC4899', '#FFF', '#7C3AED'],
  },
}
