// @famililook/shared/src/theme/haptics.js
// famililook-shared/src/theme/haptics.js
// Platform haptic language. New API (impact/pattern/setHapticsAdapter) + deprecated aliases for the 7 existing desktop2 callers.

// ──────────────────────────────────────────────────────────────
// Adapter injection (set once at app init by the native shell).
// Keeps this file buildable in a pure browser context — MUST NOT
// import @capacitor/haptics directly (breaks web builds).
// ──────────────────────────────────────────────────────────────
let _adapter = null

export function setHapticsAdapter(adapter) {
  _adapter = adapter
}

// ──────────────────────────────────────────────────────────────
// Internal helper — graceful fallback when vibration is unsupported.
// Priority: native adapter → navigator.vibrate → no-op.
// ──────────────────────────────────────────────────────────────
function vibrate(msOrPattern) {
  try {
    if (_adapter && typeof _adapter.vibrate === 'function') {
      _adapter.vibrate(msOrPattern)
      return
    }
    if (typeof navigator !== 'undefined' && typeof navigator.vibrate === 'function') {
      navigator.vibrate(msOrPattern)
    }
  } catch {
    // Intentionally silent — haptics are progressive enhancement.
    // Vibration API may be restricted on some devices/browsers.
    // Routing through AppErrorBus would risk circular dependency
    // in foundational infrastructure. This is a documented exception.
  }
}

// ──────────────────────────────────────────────────────────────
// Single impacts (all durations in milliseconds)
// ──────────────────────────────────────────────────────────────
export const impact = {
  heavy:  () => vibrate(12), // number/score lands
  medium: () => vibrate(6),  // stat or feature appears
  light:  () => vibrate(3),  // feature match highlight
  tick:   () => vibrate(1),  // icon animates in
}

// ──────────────────────────────────────────────────────────────
// Patterns (arrays passed to navigator.vibrate())
// ──────────────────────────────────────────────────────────────
export const pattern = {
  heartbeat:    () => vibrate([8, 120, 14]),          // verdict card appears
  celebration:  () => vibrate([10, 40, 10, 40, 30]),  // high score burst
  notification: () => vibrate([6, 80, 6]),            // challenge received
}

// ──────────────────────────────────────────────────────────────
// Deprecated aliases — preserve backwards compat for 7 existing
// desktop2 callers. Original desktop2 durations preserved exactly
// so haptic feel does not silently change under the alias. Remove
// in v1.0.0.
// ──────────────────────────────────────────────────────────────

// @deprecated — use impact.light() instead. Will be removed in v1.0.0
export const hapticLight = () => vibrate(10)

// @deprecated — use impact.medium() instead. Will be removed in v1.0.0
export const hapticMedium = () => vibrate(20)

// @deprecated — use pattern.heartbeat() instead. Will be removed in v1.0.0
export const hapticSuccess = () => vibrate([10, 50, 20])
