// @famililook/shared/src/theme/motion.js
// famililook-shared/src/theme/motion.js
// Motion presets extracted verbatim from famililook-desktop2/src/theme/motion.js + new framer-motion variant presets.

// ──────────────────────────────────────────────────────────────
// Spring physics (carried over from desktop2, unchanged)
// ──────────────────────────────────────────────────────────────
export const spring = {
  gentle: { type: 'spring', damping: 30, stiffness: 300 },
  snappy: { type: 'spring', damping: 22, stiffness: 380 },
  bouncy: { type: 'spring', damping: 18, stiffness: 260 },
  stiff:  { type: 'spring', damping: 40, stiffness: 500 },
}

// ──────────────────────────────────────────────────────────────
// CSS cubic-bezier easing curves
// ──────────────────────────────────────────────────────────────
export const ease = {
  spring:   'cubic-bezier(0.34, 1.56, 0.64, 1)',
  material: 'cubic-bezier(0.4, 0, 0.2, 1)',
  ios:      'cubic-bezier(0.32, 0.72, 0, 1)',
  snappy:   'cubic-bezier(0.25, 0.46, 0.45, 0.94)',
}

// ──────────────────────────────────────────────────────────────
// Stagger delays (seconds)
// ──────────────────────────────────────────────────────────────
export const stagger = {
  fast:   0.05,
  normal: 0.08,
  slow:   0.12,
}

// ──────────────────────────────────────────────────────────────
// Duration presets (seconds)
// ──────────────────────────────────────────────────────────────
export const duration = {
  instant: 0.1,
  fast:    0.2,
  normal:  0.3,
  slow:    0.5,
}

// ──────────────────────────────────────────────────────────────
// NEW — named framer-motion variant presets
// Pass as `variants` prop to motion.div. Use hidden/visible or idle/beat
// keys to drive them via `initial`/`animate` or whileInView.
// ──────────────────────────────────────────────────────────────
export const presets = {
  // Card/section enters from below, fades in
  reveal: {
    hidden:  { opacity: 0, y: 24 },
    visible: { opacity: 1, y: 0, transition: spring.gentle },
  },

  // Children stagger in sequence (use with staggerChildren on parent)
  sequence: {
    hidden:  {},
    visible: { transition: { staggerChildren: stagger.normal } },
  },

  // Number or stat slams into view (high energy)
  slam: {
    hidden:  { opacity: 0, scale: 0.6 },
    visible: { opacity: 1, scale: 1, transition: spring.bouncy },
  },

  // Gentle float in from side (used for soft moments)
  drift: {
    hidden:  { opacity: 0, x: -16 },
    visible: { opacity: 1, x: 0, transition: spring.gentle },
  },

  // Subtle breathing loop (idle state, score ring waiting)
  pulse: {
    idle: { scale: 1 },
    beat: {
      scale: 1.04,
      transition: { repeat: Infinity, repeatType: 'reverse', duration: 1.2 },
    },
  },
}
