// @famililook/shared/src/photo/index.js
// famililook-shared/src/photo/index.js
// Barrel export for Photo Quality Intelligence (Sprint X Phase X2).
// PHOTO_QUALITY_COPY constants are tested verbatim — any change breaks tests
// and requires deliberate UPL review.

export { usePhotoQuality } from './usePhotoQuality.js'
export { PhotoQualityRing } from './PhotoQualityRing.jsx'
export { useBestShotSelector } from './useBestShotSelector.js'

// ──────────────────────────────────────────────────────────────
// CEO-approved copy constants (UPL review 2026-04-18).
// Tested verbatim in tests/photo/barrel.test.js.
// Any edit requires User Psychology Lead re-review + CEO approval.
// ──────────────────────────────────────────────────────────────

export const PHOTO_QUALITY_COPY = {
  face_angle:     'A straight-on angle gives the best results',
  low_light:      'Try brighter lighting or move near a window',
  motion_blur:    'A steady hand makes all the difference',
  face_too_small: 'Move a bit closer \u2014 we need a good look at you',
  face_obscured:  'Make sure your whole face is clearly visible',
  multiple_faces: 'We analyse one face at a time \u2014 crop or retake?',
  low_resolution: 'This photo is a bit small \u2014 try a different one',
  compression:    'Use the original from your camera roll if you can',
}

// Grade-level feedback copy
export const PHOTO_QUALITY_GRADE_COPY = {
  excellent: 'Looking good',
  good:      null,  // no text — suggestion line handles it
  fair:      null,  // no text — suggestion line handles it
  poor:      'This might affect your results \u2014 but you can still continue',
}

// All 8 check keys — exported for consumer iteration and test sweeps
export const PHOTO_QUALITY_CHECKS = [
  'face_angle',
  'low_light',
  'motion_blur',
  'face_too_small',
  'face_obscured',
  'multiple_faces',
  'low_resolution',
  'compression',
]
