// @famililook/shared/src/photo/PhotoQualityRing.jsx
// famililook-shared/src/photo/PhotoQualityRing.jsx
// Animated quality ring overlay for photo previews.
// Green ring = micro-reward for good photos. Amber/red = text-only suggestion.

import React, { useState, useEffect } from 'react'
import { base } from '../theme/colors.js'

const RING_GREEN = '#30D158'
const RING_SIZE_RATIO = 1.08 // ring is 8% larger than photo
const STROKE_WIDTH = 3

/**
 * @prop {number} score - Quality score 0-100
 * @prop {string} grade - 'excellent'|'good'|'fair'|'poor'
 * @prop {string|null} suggestion - Copy string for the suggestion (already resolved from PHOTO_QUALITY_COPY)
 * @prop {boolean} loading - True while quality analysis is in progress
 * @prop {React.ReactNode} children - The photo preview element to wrap
 */
export function PhotoQualityRing({ score, grade, suggestion, loading, children }) {
  const [showGreenLabel, setShowGreenLabel] = useState(false)

  // Show "Looking good" for 1.5s when grade is excellent
  useEffect(() => {
    if (grade === 'excellent' && !loading) {
      setShowGreenLabel(true)
      const timer = setTimeout(() => setShowGreenLabel(false), 1500)
      return () => clearTimeout(timer)
    }
    setShowGreenLabel(false)
  }, [grade, loading])

  const showRing = grade === 'excellent' && !loading
  const circumference = 2 * Math.PI * 46 // r=46 in a 100x100 viewBox
  const dashOffset = circumference * (1 - score / 100)

  return (
    <div style={{ position: 'relative', display: 'inline-block' }}>
      {/* Photo content */}
      {children}

      {/* Green ring overlay — only for excellent grade */}
      {showRing && (
        <svg
          viewBox="0 0 100 100"
          style={{
            position: 'absolute',
            inset: 0,
            width: `${RING_SIZE_RATIO * 100}%`,
            height: `${RING_SIZE_RATIO * 100}%`,
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
            pointerEvents: 'none',
          }}
        >
          {/* Background track */}
          <circle
            cx="50" cy="50" r="46"
            fill="none"
            stroke="rgba(255,255,255,0.08)"
            strokeWidth={STROKE_WIDTH}
          />
          {/* Score fill */}
          <circle
            cx="50" cy="50" r="46"
            fill="none"
            stroke={RING_GREEN}
            strokeWidth={STROKE_WIDTH}
            strokeLinecap="round"
            strokeDasharray={circumference}
            strokeDashoffset={dashOffset}
            style={{
              transform: 'rotate(-90deg)',
              transformOrigin: '50% 50%',
              transition: 'stroke-dashoffset 0.6s ease-out',
            }}
          />
        </svg>
      )}

      {/* Pulse animation while loading */}
      {loading && (
        <svg
          viewBox="0 0 100 100"
          style={{
            position: 'absolute',
            inset: 0,
            width: `${RING_SIZE_RATIO * 100}%`,
            height: `${RING_SIZE_RATIO * 100}%`,
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
            pointerEvents: 'none',
            animation: 'pqr-pulse 1.2s ease-in-out infinite alternate',
          }}
        >
          <circle
            cx="50" cy="50" r="46"
            fill="none"
            stroke="rgba(255,255,255,0.15)"
            strokeWidth={STROKE_WIDTH}
          />
          <style>{`@keyframes pqr-pulse { from { opacity: 0.4; } to { opacity: 1; } }`}</style>
        </svg>
      )}

      {/* "Looking good" label — green micro-reward */}
      {showGreenLabel && (
        <div style={{
          position: 'absolute',
          bottom: -24,
          left: '50%',
          transform: 'translateX(-50%)',
          color: RING_GREEN,
          fontSize: '13px',
          fontWeight: 600,
          whiteSpace: 'nowrap',
          opacity: showGreenLabel ? 1 : 0,
          transition: 'opacity 0.2s',
        }}>
          Looking good
        </div>
      )}

      {/* Suggestion text — amber/red grades only, no ring */}
      {!loading && suggestion && grade !== 'excellent' && (
        <div style={{
          position: 'absolute',
          bottom: -24,
          left: '50%',
          transform: 'translateX(-50%)',
          color: base.textSecondary,
          fontSize: '12px',
          fontWeight: 400,
          whiteSpace: 'nowrap',
          maxWidth: '280px',
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          transition: 'opacity 0.2s',
        }}>
          {suggestion}
        </div>
      )}
    </div>
  )
}
