// @famililook/shared/src/photo/useBestShotSelector.js
// famililook-shared/src/photo/useBestShotSelector.js
// Multi-photo ranker. Runs usePhotoQuality on 2-5 files in parallel,
// returns them sorted best-first.

import { useState, useEffect, useRef } from 'react'

/**
 * @param {File[]} files - 2 to 5 photos to rank
 * @param {Object} [options]
 * @param {Function|null} [options.faceDetector=null] - Same adapter as usePhotoQuality
 * @param {Function|null} [options.analysePhoto=null] - Quality analysis function.
 *   Signature: (source, faceDetector) => Promise<{score, grade, issues, suggestion, confidence}>
 *   Injected by the barrel export to avoid circular dependency.
 * @returns {{
 *   ranked: Array<{file: File, score: number, grade: string, issues: string[], rejectionReason: string|null}>,
 *   best: File|null,
 *   loading: boolean,
 *   error: string|null,
 * }}
 */
export function useBestShotSelector(files, { faceDetector = null, analysePhoto = null } = {}) {
  const [result, setResult] = useState({
    ranked: [],
    best: null,
    loading: false,
    error: null,
  })
  const runIdRef = useRef(0)

  useEffect(() => {
    if (!files || !Array.isArray(files) || files.length === 0) return

    if (files.length < 2 || files.length > 5) {
      setResult({
        ranked: [],
        best: null,
        loading: false,
        error: 'useBestShotSelector requires 2\u20135 files',
      })
      return
    }

    if (!analysePhoto) {
      setResult({
        ranked: [],
        best: null,
        loading: false,
        error: 'analysePhoto function required \u2014 import useBestShotSelector from the barrel',
      })
      return
    }

    const runId = ++runIdRef.current
    setResult(prev => ({ ...prev, loading: true, error: null }))

    Promise.all(
      files.map(file =>
        analysePhoto(file, faceDetector)
          .then(res => ({ file, ...res, rejectionReason: res.score < 40 ? rejectionMessage(res) : null }))
          .catch(() => ({ file, score: 0, grade: 'poor', issues: ['analysis_failed'], suggestion: null, confidence: 0, rejectionReason: 'Could not analyse this photo' }))
      )
    ).then(results => {
      if (runId !== runIdRef.current) return
      const ranked = results.sort((a, b) => b.score - a.score)
      setResult({
        ranked,
        best: ranked[0]?.file || null,
        loading: false,
        error: null,
      })
    })
  }, [files, faceDetector, analysePhoto])

  return result
}

function rejectionMessage(res) {
  if (res.issues.includes('face_obscured')) return 'Face not clearly visible'
  if (res.issues.includes('multiple_faces')) return 'Multiple faces detected'
  if (res.issues.includes('motion_blur')) return 'Photo is blurry'
  if (res.issues.includes('low_light')) return 'Photo is too dark'
  if (res.issues.includes('low_resolution')) return 'Photo is too small'
  return 'Photo quality is too low'
}
