// @famililook/shared/src/photo/usePhotoQuality.js
// famililook-shared/src/photo/usePhotoQuality.js
// Client-side photo quality scoring. Accepts an optional faceDetector adapter
// to avoid duplicate TensorFlow model loading (P1 TDZ incident 2026-04-06).

import { useState, useEffect, useRef } from 'react'

/**
 * @typedef {Object} FaceDetectorResult
 * @property {number} faces - Number of faces detected
 * @property {Array<{x: number, y: number, width: number, height: number}>} boxes - Bounding boxes
 */

/**
 * Face detector adapter interface.
 * Consumer repos inject their own detector to avoid duplicate TF.js model loads.
 *
 * @callback FaceDetectorAdapter
 * @param {Blob|HTMLImageElement} source - Image to analyse
 * @returns {Promise<FaceDetectorResult>}
 *
 * @example
 * // desktop2 — passes existing detectFaces singleton:
 * import { detectFaces } from '../utils/detectFaces'
 * const quality = usePhotoQuality(file, { faceDetector: detectFaces })
 *
 * @example
 * // desktop6 — no face detection yet, passes nothing:
 * const quality = usePhotoQuality(file)
 * // Face checks skipped, canvas-based checks still run, confidence reduced.
 */

// Check priority order — most critical first
const CHECK_PRIORITY = [
  'face_obscured',
  'multiple_faces',
  'motion_blur',
  'face_angle',
  'low_light',
  'face_too_small',
  'low_resolution',
  'compression',
]

// Checks that require a faceDetector adapter
const FACE_CHECKS = new Set(['face_too_small', 'multiple_faces', 'face_angle', 'face_obscured'])

/**
 * Client-side photo quality analysis hook.
 *
 * @param {File|HTMLImageElement|HTMLCanvasElement|null} source - Photo to analyse
 * @param {Object} [options]
 * @param {FaceDetectorAdapter|null} [options.faceDetector=null] - Injected face detector.
 *   When null, face-dependent checks are skipped and confidence is reduced.
 * @returns {{
 *   score: number,
 *   grade: 'excellent'|'good'|'fair'|'poor',
 *   issues: string[],
 *   suggestion: string|null,
 *   confidence: number,
 *   loading: boolean,
 *   error: string|null,
 * }}
 */
export function usePhotoQuality(source, { faceDetector = null } = {}) {
  const [result, setResult] = useState({
    score: 0,
    grade: 'poor',
    issues: [],
    suggestion: null,
    confidence: 0,
    loading: false,
    error: null,
  })
  const runIdRef = useRef(0)

  useEffect(() => {
    if (!source) return

    const runId = ++runIdRef.current
    setResult(prev => ({ ...prev, loading: true, error: null }))

    analysePhoto(source, faceDetector).then(res => {
      if (runId !== runIdRef.current) return // stale
      setResult({ ...res, loading: false, error: null })
    }).catch(err => {
      if (runId !== runIdRef.current) return
      setResult(prev => ({ ...prev, loading: false, error: err?.message || 'Analysis failed' }))
    })
  }, [source, faceDetector])

  return result
}

// ── Internal analysis pipeline ──────────────────────────────────

async function analysePhoto(source, faceDetector) {
  const img = await loadImage(source)
  const canvas = imageToCanvas(img)
  const ctx = canvas.getContext('2d', { willReadFrequently: true })
  const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height)

  const issues = []
  let totalChecks = CHECK_PRIORITY.length
  let skippedChecks = 0

  // 1. Resolution check (synchronous)
  if (Math.min(canvas.width, canvas.height) < 480) {
    issues.push('low_resolution')
  }

  // 2. Face detection checks (if adapter provided)
  if (faceDetector) {
    try {
      const faceResult = await faceDetector(source)

      if (faceResult.faces > 1) issues.push('multiple_faces')
      if (faceResult.faces === 1 && faceResult.boxes?.[0]) {
        const box = faceResult.boxes[0]
        const faceArea = box.width * box.height
        const imageArea = canvas.width * canvas.height
        if (faceArea / imageArea < 0.15) issues.push('face_too_small')
      }
      if (faceResult.faces === 0) issues.push('face_obscured')
    } catch {
      // Face detection failed — skip face checks, reduce confidence
      skippedChecks += FACE_CHECKS.size
    }
  } else {
    skippedChecks += FACE_CHECKS.size
  }

  // 3. Canvas-based checks on full image
  const brightness = computeMeanBrightness(imageData)
  if (brightness < 80) issues.push('low_light')

  const blurScore = computeLaplacianVariance(imageData, canvas.width, canvas.height)
  if (blurScore < 100) issues.push('motion_blur')

  const dctVariance = computeDCTVariance(imageData, canvas.width, canvas.height)
  if (dctVariance < 15) issues.push('compression')

  // 4. Score computation
  const checksRun = totalChecks - skippedChecks
  const checksPassed = checksRun - issues.length
  const score = checksRun > 0 ? Math.round((checksPassed / checksRun) * 100) : 0
  const confidence = checksRun / totalChecks
  const grade = score >= 80 ? 'excellent' : score >= 60 ? 'good' : score >= 40 ? 'fair' : 'poor'

  // 5. Single suggestion — highest-priority failing check
  const suggestion = CHECK_PRIORITY.find(key => issues.includes(key)) || null

  return { score, grade, issues, suggestion, confidence }
}

// ── Image loading ───────────────────────────────────────────────

function loadImage(source) {
  return new Promise((resolve, reject) => {
    if (source instanceof HTMLImageElement) {
      if (source.complete) return resolve(source)
      source.onload = () => resolve(source)
      source.onerror = reject
      return
    }
    if (source instanceof HTMLCanvasElement) return resolve(source)

    // File or Blob
    const img = new Image()
    img.onload = () => resolve(img)
    img.onerror = reject
    img.src = URL.createObjectURL(source)
  })
}

function imageToCanvas(img) {
  const canvas = document.createElement('canvas')
  const w = img.naturalWidth || img.width
  const h = img.naturalHeight || img.height
  canvas.width = w
  canvas.height = h
  canvas.getContext('2d').drawImage(img, 0, 0)
  return canvas
}

// ── Quality check implementations ───────────────────────────────

function computeMeanBrightness(imageData) {
  const data = imageData.data
  let sum = 0
  const pixelCount = data.length / 4
  for (let i = 0; i < data.length; i += 4) {
    sum += (data[i] + data[i + 1] + data[i + 2]) / 3
  }
  return sum / pixelCount
}

function computeLaplacianVariance(imageData, width, height) {
  const grey = toGreyscale(imageData)
  // Laplacian kernel [0,1,0,1,-4,1,0,1,0]
  let sum = 0
  let sumSq = 0
  let count = 0
  for (let y = 1; y < height - 1; y++) {
    for (let x = 1; x < width - 1; x++) {
      const idx = y * width + x
      const lap = grey[idx - width] + grey[idx + width] +
                  grey[idx - 1] + grey[idx + 1] - 4 * grey[idx]
      sum += lap
      sumSq += lap * lap
      count++
    }
  }
  const mean = sum / count
  return (sumSq / count) - (mean * mean)
}

function computeDCTVariance(imageData, width, height) {
  const grey = toGreyscale(imageData)
  // Sample 8x8 blocks across the image
  const blockSize = 8
  const stepsX = Math.floor(width / blockSize)
  const stepsY = Math.floor(height / blockSize)
  if (stepsX < 2 || stepsY < 2) return 100 // too small to assess

  let totalVariance = 0
  let blockCount = 0
  // Sample up to 64 blocks evenly distributed
  const skipX = Math.max(1, Math.floor(stepsX / 8))
  const skipY = Math.max(1, Math.floor(stepsY / 8))

  for (let by = 0; by < stepsY; by += skipY) {
    for (let bx = 0; bx < stepsX; bx += skipX) {
      let bSum = 0
      let bSumSq = 0
      for (let dy = 0; dy < blockSize; dy++) {
        for (let dx = 0; dx < blockSize; dx++) {
          const px = grey[(by * blockSize + dy) * width + (bx * blockSize + dx)]
          bSum += px
          bSumSq += px * px
        }
      }
      const n = blockSize * blockSize
      const bMean = bSum / n
      totalVariance += (bSumSq / n) - (bMean * bMean)
      blockCount++
    }
  }
  return blockCount > 0 ? totalVariance / blockCount : 100
}

function toGreyscale(imageData) {
  const data = imageData.data
  const grey = new Float32Array(data.length / 4)
  for (let i = 0; i < grey.length; i++) {
    const j = i * 4
    grey[i] = 0.299 * data[j] + 0.587 * data[j + 1] + 0.114 * data[j + 2]
  }
  return grey
}
