// @famililook/shared/src/rewards/MicroReward.jsx
// famililook-shared/src/rewards/MicroReward.jsx
// Generic reward moment wrapper. Maps reward types to colours, haptics, and animations.
// Milestone uses pattern.celebration (not heartbeat — UPL review 2026-04-18).

import React, { useEffect, useRef } from 'react'
import { reward } from '../theme/dopamine.js'
import { impact, pattern } from '../theme/haptics.js'
import { presets } from '../theme/motion.js'

const REWARD_CONFIG = {
  achievement: {
    color: reward.achievement,   // #F5A623
    haptic: impact.heavy,
    preset: presets.slam,
  },
  rare: {
    color: reward.anticipation,  // #7C3AED
    haptic: impact.medium,
    preset: presets.drift,
  },
  milestone: {
    color: reward.electric,      // #0A84FF
    haptic: pattern.celebration, // UPL: pattern.celebration, NOT heartbeat
    preset: presets.reveal,
  },
  social: {
    color: reward.social,        // #EC4899
    haptic: impact.light,
    preset: presets.reveal,
  },
}

/**
 * @prop {'achievement'|'rare'|'milestone'|'social'} type
 * @prop {React.ReactNode} children - The content to reward
 * @prop {boolean} active - Set true to trigger reward animation
 * @prop {number} delay - ms before animation fires (default 0)
 */
export function MicroReward({ type, children, active = false, delay = 0 }) {
  const config = REWARD_CONFIG[type] || REWARD_CONFIG.achievement
  const hasFiredRef = useRef(false)

  useEffect(() => {
    if (active && !hasFiredRef.current) {
      hasFiredRef.current = true
      if (delay > 0) {
        const timer = setTimeout(() => config.haptic(), delay)
        return () => clearTimeout(timer)
      }
      config.haptic()
    }
    if (!active) {
      hasFiredRef.current = false
    }
  }, [active, config, delay])

  const style = {
    transition: `opacity ${delay > 0 ? delay + 300 : 300}ms ease-out, transform 300ms ease-out`,
    opacity: active ? 1 : 0,
    transform: active ? 'scale(1) translateY(0)' : scaleForPreset(config.preset),
  }

  return (
    <div style={style}>
      {children}
    </div>
  )
}

function scaleForPreset(preset) {
  if (preset === presets.slam) return 'scale(0.6)'
  if (preset === presets.drift) return 'translateX(-16px)'
  return 'translateY(24px)'
}

// Export config for testing
export { REWARD_CONFIG }
