// @famililook/shared/src/rewards/ProgressReveal.jsx
// famililook-shared/src/rewards/ProgressReveal.jsx
// 8-icon animated loading state with accelerating timing curve.
// UPL-approved: 500/375/250ms per icon, 50ms cascade fast-forward.

import React, { useState, useEffect, useRef } from 'react'
import { products } from '../theme/colors.js'
import { impact } from '../theme/haptics.js'
import { featureIconMap } from '../icons/index.js'

const FEATURE_KEYS = ['eyes', 'eyebrows', 'smile', 'nose', 'face_shape', 'skin', 'hair', 'ears']

// Accelerating timing curve (UPL: builds momentum, feels faster)
// Icons 1-3: 500ms, Icons 4-6: 375ms, Icons 7-8: 250ms
const ICON_DELAYS = [500, 500, 500, 375, 375, 375, 250, 250]

const PRODUCT_COPY = {
  famililook:  'features',
  familimatch: 'compatibility',
}

/**
 * @prop {string} productId - 'famililook' | 'familimatch' — determines icon colour + copy
 * @prop {number} estimatedMs - Estimated total analysis time (unused — timing is curve-based)
 * @prop {boolean} complete - Set to true when analysis resolves
 * @prop {function} onComplete - Called after exit animation finishes
 */
export function ProgressReveal({ productId = 'famililook', estimatedMs = 3000, complete = false, onComplete }) {
  const [litCount, setLitCount] = useState(0)
  const timerRef = useRef(null)
  const completeCalledRef = useRef(false)

  const accentColor = products[productId]?.primary || products.famililook.primary
  const analysingWord = PRODUCT_COPY[productId] || 'features'

  // Sequential icon lighting with accelerating curve
  useEffect(() => {
    if (complete) return // handled by fast-forward effect below

    if (litCount >= 8) return // all lit, wait for complete

    const delay = ICON_DELAYS[litCount] || 250
    timerRef.current = setTimeout(() => {
      setLitCount(prev => prev + 1)
      impact.tick()
    }, delay)

    return () => clearTimeout(timerRef.current)
  }, [litCount, complete])

  // Fast-forward when analysis completes before all 8 lit
  useEffect(() => {
    if (!complete) return
    if (completeCalledRef.current) return

    if (litCount >= 8) {
      completeCalledRef.current = true
      onComplete?.()
      return
    }

    // Cascade remaining icons at 50ms intervals
    const remaining = 8 - litCount
    let i = 0
    const cascade = setInterval(() => {
      i++
      setLitCount(prev => prev + 1)
      impact.tick()
      if (i >= remaining) {
        clearInterval(cascade)
        setTimeout(() => {
          completeCalledRef.current = true
          onComplete?.()
        }, 100)
      }
    }, 50)

    return () => clearInterval(cascade)
  }, [complete, litCount, onComplete])

  // Pulse animation for last icon when waiting for analysis
  const isWaiting = litCount >= 8 && !complete

  return (
    <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', gap: '16px' }}>
      {/* Icon grid — 2x4 */}
      <div style={{
        display: 'grid',
        gridTemplateColumns: 'repeat(4, 1fr)',
        gap: '12px',
      }}>
        {FEATURE_KEYS.map((key, i) => {
          const Icon = featureIconMap[key]
          const isLit = i < litCount
          const isPulsing = isWaiting && i === 7

          return (
            <div
              key={key}
              style={{
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                width: '40px',
                height: '40px',
                borderRadius: '10px',
                background: isLit ? `${accentColor}1A` : 'rgba(255,255,255,0.04)',
                transition: 'opacity 0.3s, background 0.3s',
                opacity: isLit ? 1 : 0.2,
                animation: isPulsing ? 'pr-pulse 1.2s ease-in-out infinite alternate' : 'none',
              }}
            >
              {Icon && <Icon size={20} />}
            </div>
          )
        })}
      </div>

      {/* Progress text */}
      <div style={{
        fontSize: '13px',
        fontWeight: 400,
        color: 'rgba(255,255,255,0.6)',
        textAlign: 'center',
      }}>
        {litCount < 8
          ? `Analysing your ${analysingWord}\u2026 ${litCount} of 8 traits complete`
          : isWaiting
            ? `Finalising your ${analysingWord}\u2026`
            : `Analysis complete`
        }
      </div>

      <style>{`@keyframes pr-pulse { from { opacity: 0.6; } to { opacity: 1; } }`}</style>
    </div>
  )
}
