// @famililook/shared/src/rewards/StatHighlight.jsx
// famililook-shared/src/rewards/StatHighlight.jsx
// Gold-highlighted rare stat callout. Auto-dismisses after 4 seconds.
// Product-aware rarity labels per UPL review (2026-04-18).

import React, { useEffect, useState } from 'react'
import { reward } from '../theme/dopamine.js'
import { base } from '../theme/colors.js'
import { impact } from '../theme/haptics.js'

// Product-aware rarity tier labels (UPL: gaming vocab for Match, family vocab for Look)
export const RARITY_LABELS = {
  familimatch: {
    common:    'Common',
    uncommon:  'Uncommon',
    rare:      'Rare',
    epic:      'Epic',
    legendary: 'Legendary',
    mythic:    'Mythic',
  },
  famililook: {
    common:    'Typical',
    uncommon:  'Notable',
    rare:      'Unusual',
    epic:      'Distinctive',
    legendary: 'Exceptional',
    mythic:    'Extraordinary',
  },
}

const RARITY_COLORS = {
  common:    '#94A3B8',
  uncommon:  '#22D3EE',
  rare:      '#A855F7',
  epic:      '#FBBF24',
  legendary: '#FBBF24',
  mythic:    '#F0ABFC',
}

/**
 * @prop {object} reward - Single item from useVariableReward rewards array
 *   { type, label, value, rarity }
 * @prop {string} productId - 'famililook' | 'familimatch' — selects rarity label register
 * @prop {function} onDismiss - Called after auto-dismiss (4s) or swipe
 */
export function StatHighlight({ reward: rewardItem, productId = 'familimatch', onDismiss }) {
  const [visible, setVisible] = useState(true)

  useEffect(() => {
    impact.medium()

    const timer = setTimeout(() => {
      setVisible(false)
      setTimeout(() => onDismiss?.(), 200) // wait for fade-out
    }, 4000)

    return () => clearTimeout(timer)
  }, [onDismiss])

  const rarityLabel = RARITY_LABELS[productId]?.[rewardItem.rarity]
    || RARITY_LABELS.familimatch[rewardItem.rarity]
    || rewardItem.rarity

  const rarityColor = RARITY_COLORS[rewardItem.rarity] || '#94A3B8'

  // Swipe-to-dismiss
  const handleTouchStart = (e) => {
    const startY = e.touches[0].clientY
    const handleTouchEnd = (e2) => {
      const delta = Math.abs(e2.changedTouches[0].clientY - startY)
      if (delta > 40) {
        setVisible(false)
        setTimeout(() => onDismiss?.(), 200)
      }
      document.removeEventListener('touchend', handleTouchEnd)
    }
    document.addEventListener('touchend', handleTouchEnd)
  }

  return (
    <div
      onTouchStart={handleTouchStart}
      style={{
        background: `${reward.achievement}26`, // 15% opacity
        border: `1px solid ${reward.achievement}66`, // 40% opacity
        borderRadius: '12px',
        padding: '12px 16px',
        display: 'flex',
        alignItems: 'center',
        gap: '10px',
        opacity: visible ? 1 : 0,
        transform: visible ? 'translateX(0)' : 'translateX(-16px)',
        transition: 'opacity 0.2s, transform 0.2s',
      }}
    >
      {/* Rarity badge */}
      <span style={{
        fontSize: '10px',
        fontWeight: 700,
        textTransform: 'uppercase',
        letterSpacing: '0.05em',
        color: rarityColor,
        flexShrink: 0,
      }}>
        {rarityLabel}
      </span>

      {/* Label text */}
      <span style={{
        fontSize: '13px',
        fontWeight: 500,
        color: base.textPrimary,
        lineHeight: 1.3,
      }}>
        {rewardItem.label}
      </span>
    </div>
  )
}
